torchrl/envs/batched_envs.py

```diff
@@ -1328,13 +1465,18 @@ def _reset(self, tensordict: TensorDictBase, **kwargs) -> TensorDictBase:
                 (self.num_workers,), device=self.device, dtype=torch.bool
             )
 
-        workers = []
-
-        for i, channel in enumerate(self.parent_channels):
+        outs = []
+        for i in range(self.num_workers):
             if tensordict is not None:
                 tensordict_ = tensordict[i]
                 if tensordict_.is_empty():
                     tensordict_ = None
+                elif self.device is not None and self.device.type == "mps":
+                    # copy_ fails when moving mps->cpu using copy_
+                    # in some cases when a view of an mps tensor is used.
+                    # We know the shared tensors are not MPS, so we can
+                    # safely assume that the shared tensors are on cpu
+                    tensordict_ = tensordict_.to("cpu")
```

Comment:
@skandermoalla this will make the class usable, though it's not optimal.
this _reset will not be used much in practice and the info copied will usually be lightweight (e.g. "_reset" info etc) so that should not slow things down too much

Reply:
Thanks for this!